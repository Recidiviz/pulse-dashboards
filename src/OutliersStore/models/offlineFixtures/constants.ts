// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { ascending, range } from "d3-array";
import { subMonths } from "date-fns";
import { z } from "zod";

export const ADVERSE_METRIC_IDS = z.enum([
  "incarceration_starts",
  "absconsions_bench_warrants",
  "incarceration_starts_technical_violation",
]);

export const CASELOAD_TYPE_IDS = z.enum(["GENERAL_OR_OTHER", "SEX_OFFENSE"]);

export const LATEST_END_DATE = new Date(2023, 8, 1);

export const LOOKBACK_END_DATES = range(6)
  .map((offset) => subMonths(LATEST_END_DATE, offset))
  .sort(ascending);
