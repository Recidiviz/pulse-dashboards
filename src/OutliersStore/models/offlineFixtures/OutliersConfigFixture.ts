// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { OutliersConfig } from "../OutliersConfig";
import { ADVERSE_METRIC_IDS } from "./constants";

export const OutliersConfigFixture: OutliersConfig = {
  supervisionOfficerLabel: "officer",
  supervisionDistrictLabel: "region",
  supervisionDistrictManagerLabel: "district manager",
  supervisionJiiLabel: "client",
  supervisionSupervisorLabel: "supervisor",
  supervisionUnitLabel: "team",
  learnMoreUrl: "https://recidiviz.org",
  metrics: [
    {
      name: ADVERSE_METRIC_IDS.enum.absconsions_bench_warrants,
      bodyDisplayName: "absconsion rate",
      titleDisplayName: "Absconsion Rate",
      eventName: "absconsions",
      outcomeType: "ADVERSE",
    },
    {
      name: ADVERSE_METRIC_IDS.enum.incarceration_starts,
      bodyDisplayName: "incarceration rate",
      titleDisplayName: "Incarceration Rate",
      eventName: "incarcerations",
      outcomeType: "ADVERSE",
    },
    {
      name: ADVERSE_METRIC_IDS.enum.incarceration_starts_technical_violation,
      bodyDisplayName: "technical incarceration rate",
      titleDisplayName: "Technical Incarceration Rate",
      eventName: "technical incarcerations",
      outcomeType: "ADVERSE",
    },
  ],
  clientEvents: [
    {
      displayName: "Violations",
      name: "violations",
    },
    {
      displayName: "Sanctions",
      name: "violation_responses",
    },
  ],
};
