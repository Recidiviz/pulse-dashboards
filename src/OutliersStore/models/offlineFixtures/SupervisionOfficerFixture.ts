// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import {
  RawSupervisionOfficer,
  supervisionOfficerSchema,
} from "../SupervisionOfficer";
import { ADVERSE_METRIC_IDS, CASELOAD_TYPE_IDS } from "./constants";
import { rawSupervisionOfficerMetricOutlierFixtures } from "./SupervisionOfficerMetricOutlierFixture";
import { supervisionOfficerSupervisorsFixture } from "./SupervisionOfficerSupervisor";

export const rawSupervisionOfficerFixture: RawSupervisionOfficer[] = [
  {
    externalId: "so1",
    fullName: {
      given_names: "Duke",
      surname: "Ellington",
    },
    district: supervisionOfficerSupervisorsFixture[0].district,
    supervisorId: supervisionOfficerSupervisorsFixture[0].externalId,
    caseloadTypes: [
      CASELOAD_TYPE_IDS.enum.GENERAL_OR_OTHER,
      CASELOAD_TYPE_IDS.enum.SEX_OFFENSE,
    ],
    currentPeriodStatuses: {
      far: [
        rawSupervisionOfficerMetricOutlierFixtures
          .absconsions_bench_warrants[0],
        rawSupervisionOfficerMetricOutlierFixtures.incarceration_starts[0],
      ],
      near: [ADVERSE_METRIC_IDS.enum.incarceration_starts_technical_violation],
      met: [],
    },
  },
  {
    externalId: "so2",
    fullName: {
      given_names: "Chet",
      surname: "Baker",
    },
    district: supervisionOfficerSupervisorsFixture[0].district,
    supervisorId: supervisionOfficerSupervisorsFixture[0].externalId,
    caseloadTypes: [CASELOAD_TYPE_IDS.enum.GENERAL_OR_OTHER],
    currentPeriodStatuses: {
      far: [],
      near: [
        ADVERSE_METRIC_IDS.enum.incarceration_starts,
        ADVERSE_METRIC_IDS.enum.incarceration_starts_technical_violation,
      ],
      met: [ADVERSE_METRIC_IDS.enum.absconsions_bench_warrants],
    },
  },
  {
    externalId: "so3",
    fullName: {
      given_names: "Louis",
      surname: "Armstrong",
    },
    district: supervisionOfficerSupervisorsFixture[0].district,
    supervisorId: supervisionOfficerSupervisorsFixture[0].externalId,
    caseloadTypes: [CASELOAD_TYPE_IDS.enum.GENERAL_OR_OTHER],
    currentPeriodStatuses: {
      far: [
        rawSupervisionOfficerMetricOutlierFixtures
          .absconsions_bench_warrants[1],
      ],
      near: [ADVERSE_METRIC_IDS.enum.incarceration_starts],
      met: [ADVERSE_METRIC_IDS.enum.incarceration_starts_technical_violation],
    },
  },
  {
    externalId: "so4",
    fullName: {
      given_names: "William James",
      middle_names: '"Count"',
      surname: "Basie",
    },
    district: supervisionOfficerSupervisorsFixture[1].district,
    supervisorId: supervisionOfficerSupervisorsFixture[1].externalId,
    caseloadTypes: [CASELOAD_TYPE_IDS.enum.GENERAL_OR_OTHER],
    currentPeriodStatuses: {
      far: [
        rawSupervisionOfficerMetricOutlierFixtures
          .absconsions_bench_warrants[2],
      ],
      near: [ADVERSE_METRIC_IDS.enum.incarceration_starts],
      met: [ADVERSE_METRIC_IDS.enum.incarceration_starts_technical_violation],
    },
  },
];

export const supervisionOfficerFixture = rawSupervisionOfficerFixture.map(
  (officer) => supervisionOfficerSchema.parse(officer)
);
