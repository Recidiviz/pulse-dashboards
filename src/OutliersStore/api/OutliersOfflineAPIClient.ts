/* eslint-disable class-methods-use-this */
// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { differenceInDays, subDays } from "date-fns";

import { ClientEvent } from "../models/ClientEvent";
import { ClientInfo } from "../models/ClientInfo";
import { MetricBenchmark } from "../models/MetricBenchmark";
import { LATEST_END_DATE } from "../models/offlineFixtures/constants";
import { SupervisionOfficer } from "../models/SupervisionOfficer";
import { SupervisionOfficerMetricEvent } from "../models/SupervisionOfficerMetricEvent";
import { SupervisionOfficerSupervisor } from "../models/SupervisionOfficerSupervisor";
import type { OutliersStore } from "../OutliersStore";
import { OutliersAPI } from "./interface";

export class OutliersOfflineAPIClient implements OutliersAPI {
  // eslint-disable-next-line no-useless-constructor
  constructor(public readonly outliersStore: OutliersStore) {}

  async init() {
    const config = (
      await import("../models/offlineFixtures/OutliersConfigFixture")
    ).OutliersConfigFixture;

    return config;
  }

  async metricBenchmarks(): Promise<MetricBenchmark[]> {
    const { metricBenchmarksFixture } = await import(
      "../models/offlineFixtures/MetricBenchmarkFixture"
    );
    return metricBenchmarksFixture;
  }

  async supervisionOfficerSupervisors(): Promise<
    SupervisionOfficerSupervisor[]
  > {
    const { supervisionOfficerSupervisorsFixture } = await import(
      "../models/offlineFixtures/SupervisionOfficerSupervisor"
    );
    return supervisionOfficerSupervisorsFixture;
  }

  async officersForSupervisor(
    supervisorPseudoId: string
  ): Promise<Array<SupervisionOfficer>> {
    const { supervisionOfficerFixture } = await import(
      "../models/offlineFixtures/SupervisionOfficerFixture"
    );

    return supervisionOfficerFixture.filter(
      (o) => `hashed-${o.supervisorExternalId}` === supervisorPseudoId
    );
  }

  async supervisionOfficer(
    officerPseudoId: string
  ): Promise<SupervisionOfficer> {
    const { supervisionOfficerFixture } = await import(
      "../models/offlineFixtures/SupervisionOfficerFixture"
    );

    const officerFixture = supervisionOfficerFixture.find(
      (o) => o.pseudonymizedId === officerPseudoId
    );

    if (!officerFixture)
      throw new Error(`Officer ${officerPseudoId} not present in fixture data`);

    return officerFixture;
  }

  async supervisionOfficerMetricEvents(
    officerPseudoId: string,
    metricId: string
  ): Promise<SupervisionOfficerMetricEvent[]> {
    const { supervisionOfficerMetricEventFixture } = await import(
      "../models/offlineFixtures/SupervisionOfficerMetricEventFixture"
    );
    return supervisionOfficerMetricEventFixture.filter(
      (e) => e.metricId === metricId
    );
  }

  async clientInfo(clientPseudoId: string): Promise<ClientInfo> {
    const { clientInfoFixture } = await import(
      "../models/offlineFixtures/ClientInfoFixture"
    );

    const clientInfo = clientInfoFixture[clientPseudoId];

    if (!clientInfo) {
      throw new Error(`Client ${clientPseudoId} not present in fixture data`);
    }

    return clientInfo;
  }

  async clientEvents(
    clientPseudoId: string,
    endDate: Date
  ): Promise<Array<ClientEvent>> {
    const { clientEventFixture } = await import(
      "../models/offlineFixtures/ClientEventFixture"
    );

    // adjust fixture dates relative to the input
    const dateOffsetInDays = differenceInDays(LATEST_END_DATE, endDate);

    return clientEventFixture.map((event) => ({
      ...event,
      eventDate: subDays(event.eventDate, dateOffsetInDays),
    }));
  }
}
