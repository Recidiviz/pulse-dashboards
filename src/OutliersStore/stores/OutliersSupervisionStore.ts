// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { index } from "d3-array";
import { makeAutoObservable, observable } from "mobx";

import { isOfflineMode } from "../../utils/isOfflineMode";
import { OutliersAPI } from "../api/interface";
import { MetricBenchmark } from "../models/MetricBenchmark";
import { MetricConfig } from "../models/MetricConfig";
import { OutliersConfig } from "../models/OutliersConfig";
import { SupervisionOfficer } from "../models/SupervisionOfficer";
import { SupervisionOfficerMetricEvent } from "../models/SupervisionOfficerMetricEvent";
import { SupervisionOfficerSupervisor } from "../models/SupervisionOfficerSupervisor";
import type { OutliersStore } from "../OutliersStore";
import { FlowMethod, StringMap2D } from "../types";

export class OutliersSupervisionStore {
  private benchmarksByMetricAndCaseloadType?: StringMap2D<MetricBenchmark>;

  officersBySupervisorPseudoId: Map<string, SupervisionOfficer[]> = new Map();

  supervisorPseudoId?: string;

  officerPseudoId?: string;

  metricId?: string;

  private allSupervisionOfficerSupervisors?: SupervisionOfficerSupervisor[];

  metricEventsByOfficerPseudoIdAndMetricId: StringMap2D<
    Array<SupervisionOfficerMetricEvent>
  > = new Map();

  constructor(
    public readonly outliersStore: OutliersStore,
    private readonly config: OutliersConfig
  ) {
    makeAutoObservable<this, "config">(this, {
      // this object will be static so there's no need to deeply observe it
      config: observable.ref,
    });
  }

  /**
   * Fetches metric benchmark data for the current tenant.
   *
   * This is a MobX flow method and should be called with mobx.flowResult.
   */
  *hydrateMetricConfigs(): FlowMethod<OutliersAPI["metricBenchmarks"], void> {
    if (this.benchmarksByMetricAndCaseloadType) return;

    const benchmarks = yield this.outliersStore.apiClient.metricBenchmarks();
    const benchmarksByMetricAndCaseloadType = index(
      benchmarks,
      (b) => b.metricId,
      (b) => b.caseloadType
    );

    this.benchmarksByMetricAndCaseloadType = benchmarksByMetricAndCaseloadType;
  }

  /**
   * Provides a mapping of all configured metrics, including nested benchmarks data.
   * If undefined, it is still awaiting hydration.
   */
  get metricConfigsById(): Map<string, MetricConfig> | undefined {
    const { benchmarksByMetricAndCaseloadType } = this;
    if (!benchmarksByMetricAndCaseloadType) return;

    return index(
      this.config.metrics.map((m): MetricConfig => {
        const metricBenchmarksByCaseloadType =
          benchmarksByMetricAndCaseloadType.get(m.name);

        return {
          ...m,
          metricBenchmarksByCaseloadType,
        };
      }),
      (m) => m.name
    );
  }

  /**
   * Provides a mapping of all configured adverse outcome metrics, including nested benchmarks data.
   */
  get adverseMetricConfigsById(): Map<string, MetricConfig> | undefined {
    if (!this.metricConfigsById) return;

    return new Map(
      Array.from(this.metricConfigsById.entries()).filter(
        ([id, m]) => m.outcomeType === "ADVERSE"
      )
    );
  }

  get currentSupervisorUser(): SupervisionOfficerSupervisor | undefined {
    const { userAppMetadata } = this.outliersStore.rootStore.userStore;
    if (isOfflineMode()) return undefined;

    if (userAppMetadata) {
      const { role, externalId, district, pseudonymizedId } = userAppMetadata;
      if (
        // until metadata is extended with role subtypes,
        // this should be a safe proxy for distinguishing supervisors from leadership
        // (as long as no other line staff are granted access)
        role === "supervision_staff" &&
        // nothing will work without these ID values, so we can't construct a useful
        // supervisor identity record if we didn't get them both from Auth0
        externalId &&
        pseudonymizedId
      ) {
        return {
          externalId,
          supervisionDistrict: district ?? null,
          pseudonymizedId,
          // unfortunately we don't reliably get anyone's name from auth0
          fullName: {},
          displayName: "",
        };
      }
    }
  }

  get supervisionOfficerSupervisors():
    | SupervisionOfficerSupervisor[]
    | undefined {
    // if the user is a supervisor, the only record they need is their own
    if (this.currentSupervisorUser) return [this.currentSupervisorUser];

    return this.allSupervisionOfficerSupervisors;
  }

  supervisionOfficerSupervisorByExternalId(
    supervisorId: string
  ): SupervisionOfficerSupervisor | undefined {
    return this.supervisionOfficerSupervisors?.find(
      (s) => s.externalId === supervisorId
    );
  }

  supervisionOfficerSupervisorByPseudoId(
    supervisorPseudoId: string
  ): SupervisionOfficerSupervisor | undefined {
    return this.supervisionOfficerSupervisors?.find(
      (s) => s.pseudonymizedId === supervisorPseudoId
    );
  }

  /**
   * Fetches supervision officer supervisor data for the current tenant.
   */
  *hydrateSupervisionOfficerSupervisors(): FlowMethod<
    OutliersAPI["supervisionOfficerSupervisors"],
    void
  > {
    // note that this will bypass hydration if the current user is a supervisor
    // (since we expect to have already populated this list with their Auth0 data).
    // we expect the API request to fail for these users anyway so there is no reason
    // to let the request proceed
    if (this.supervisionOfficerSupervisors) return;

    this.allSupervisionOfficerSupervisors =
      yield this.outliersStore.apiClient.supervisionOfficerSupervisors();
  }

  /**
   * Fetches officer and metric data for the specified supervisor
   */
  *hydrateOfficersForSupervisor(
    supervisorPseudoId: string
  ): FlowMethod<OutliersAPI["officersForSupervisor"], void> {
    if (this.officersBySupervisorPseudoId.has(supervisorPseudoId)) return;

    const officersData =
      yield this.outliersStore.apiClient.officersForSupervisor(
        supervisorPseudoId
      );

    if (officersData.length > 0)
      this.officersBySupervisorPseudoId.set(supervisorPseudoId, officersData);
  }

  setSupervisorPseudoId(supervisorPseudoId: string | undefined): void {
    this.supervisorPseudoId = supervisorPseudoId;
  }

  setOfficerPseudoId(officerPseudoId: string | undefined): void {
    this.officerPseudoId = officerPseudoId;
  }

  setMetricId(metricId: string | undefined): void {
    this.metricId = metricId;
  }

  /*
   * Fetches events data for the specified officer and metric.
   */
  *hydrateMetricEventsForOfficer(
    officerPseudoId: string,
    metricId: string
  ): FlowMethod<OutliersAPI["supervisionOfficerMetricEvents"], void> {
    if (
      this.metricEventsByOfficerPseudoIdAndMetricId
        .get(officerPseudoId)
        ?.has(metricId)
    )
      return;

    const eventsData =
      yield this.outliersStore.apiClient.supervisionOfficerMetricEvents(
        officerPseudoId,
        metricId
      );

    const metricsMap =
      this.metricEventsByOfficerPseudoIdAndMetricId.get(officerPseudoId) ??
      new Map();
    metricsMap.set(metricId, eventsData);
    if (!this.metricEventsByOfficerPseudoIdAndMetricId.has(officerPseudoId)) {
      this.metricEventsByOfficerPseudoIdAndMetricId.set(
        officerPseudoId,
        metricsMap
      );
    }
  }
}
