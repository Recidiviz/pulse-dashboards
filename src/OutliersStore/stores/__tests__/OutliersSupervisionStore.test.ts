// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { cloneDeep } from "lodash";
import { configure, flowResult } from "mobx";
import { ValuesType } from "utility-types";

import { RootStore } from "../../../RootStore";
import { isOfflineMode } from "../../../utils/isOfflineMode";
import { OutliersOfflineAPIClient } from "../../api/OutliersOfflineAPIClient";
import { CASELOAD_TYPE_IDS } from "../../models/offlineFixtures/constants";
import { metricBenchmarksFixture } from "../../models/offlineFixtures/MetricBenchmarkFixture";
import { OutliersConfigFixture } from "../../models/offlineFixtures/OutliersConfigFixture";
import { supervisionOfficerFixture } from "../../models/offlineFixtures/SupervisionOfficerFixture";
import { supervisionOfficerSupervisorsFixture } from "../../models/offlineFixtures/SupervisionOfficerSupervisor";
import { OutliersConfig } from "../../models/OutliersConfig";
import { OutliersStore } from "../../OutliersStore";
import { OutliersSupervisionStore } from "../OutliersSupervisionStore";

let store: OutliersSupervisionStore;

beforeEach(() => {
  jest.resetModules();
  configure({ safeDescriptors: false });
  store = new OutliersSupervisionStore(
    new OutliersStore(new RootStore()),
    OutliersConfigFixture
  );
});

afterEach(() => {
  jest.restoreAllMocks();
  configure({ safeDescriptors: true });
});

test("hydrate benchmarks", async () => {
  await expect(flowResult(store.hydrateMetricConfigs())).toResolve();
});

test("cannot hydrate benchmarks with API error", async () => {
  jest
    .spyOn(OutliersOfflineAPIClient.prototype, "metricBenchmarks")
    .mockRejectedValue(new Error("something went wrong"));

  await expect(flowResult(store.hydrateMetricConfigs())).toReject();

  expect(store.metricConfigsById).toBeUndefined();
});

// The verification for missing metrics happens in the presenter where we
// have a more accurate idea of which metrics are necessary per supervisor/caseload type
test("can hydrate benchmarks with missing metrics", async () => {
  jest
    .spyOn(OutliersOfflineAPIClient.prototype, "metricBenchmarks")
    // this should include only 2 of the 3 expected metric types
    .mockResolvedValue(metricBenchmarksFixture.slice(0, 2));

  await expect(flowResult(store.hydrateMetricConfigs())).toResolve();

  expect(store.metricConfigsById).toBeUndefined();
});

test("configs include benchmarks", async () => {
  await flowResult(store.hydrateMetricConfigs());

  store.metricConfigsById?.forEach((mc) => {
    // benchmarks data should be hydrated
    expect(mc.metricBenchmarksByCaseloadType).toBeDefined();
    // both caseload types should be present for all
    expect(Array.from(mc.metricBenchmarksByCaseloadType?.keys() ?? [])).toEqual(
      expect.arrayContaining(CASELOAD_TYPE_IDS.options)
    );
  });

  expect.assertions(OutliersConfigFixture.metrics.length * 2);
});

test("hydrated benchmarks can be missing caseload types", async () => {
  jest
    .spyOn(OutliersOfflineAPIClient.prototype, "metricBenchmarks")
    // this should be missing the SEX_OFFENSE caseload type for all metrics
    .mockResolvedValue(metricBenchmarksFixture.slice(0, 3));

  await expect(flowResult(store.hydrateMetricConfigs())).toResolve();

  expect(store.metricConfigsById).toBeDefined();

  store.metricConfigsById?.forEach((mc) => {
    expect(
      mc.metricBenchmarksByCaseloadType?.get(
        CASELOAD_TYPE_IDS.enum.GENERAL_OR_OTHER
      )
    ).toBeDefined();
    expect(
      mc.metricBenchmarksByCaseloadType?.get(CASELOAD_TYPE_IDS.enum.SEX_OFFENSE)
    ).toBeUndefined();
  });
});

test("adverse metric configs", async () => {
  const additionalConfigsFixture = cloneDeep(OutliersConfigFixture);
  const favorableMetricConfig: ValuesType<OutliersConfig["metrics"]> = {
    name: "transfers_to_liberty",
    bodyDisplayName: "successful discharge rate",
    titleDisplayName: "Successful discharge rate",
    eventName: "successful discharges",
    outcomeType: "FAVORABLE",
  };
  additionalConfigsFixture.metrics.push(favorableMetricConfig);

  store = new OutliersSupervisionStore(
    new OutliersStore(new RootStore()),
    additionalConfigsFixture
  );

  expect(store.adverseMetricConfigsById).toBeUndefined();

  // also need to hydrate benchmark data.
  // This is not realistic for this metric but it doesn't really matter for this test
  const additionalMetricBenchmark = cloneDeep(metricBenchmarksFixture[0]);
  additionalMetricBenchmark.metricId = favorableMetricConfig.name;
  jest
    .spyOn(OutliersOfflineAPIClient.prototype, "metricBenchmarks")
    .mockResolvedValue([...metricBenchmarksFixture, additionalMetricBenchmark]);

  await flowResult(store.hydrateMetricConfigs());

  // base config mapping should have all metrics
  additionalConfigsFixture.metrics.forEach((m) => {
    expect(store.metricConfigsById?.get(m.name)).toBeDefined();
  });

  // adverse config mapping should not include the additional one we injected
  OutliersConfigFixture.metrics.forEach((m) => {
    expect(store.adverseMetricConfigsById?.get(m.name)).toBeDefined();
  });
  expect(
    store.adverseMetricConfigsById?.get(favorableMetricConfig.name)
  ).toBeUndefined();
});

test("hydrate supervisionOfficerSupervisors", async () => {
  expect(store.supervisionOfficerSupervisors).toBeFalsy();

  await expect(
    flowResult(store.hydrateSupervisionOfficerSupervisors())
  ).resolves.not.toThrow();
  expect(store.supervisionOfficerSupervisors?.length).toBe(
    supervisionOfficerSupervisorsFixture.length
  );
});

test("hydrate supervisionOfficers for supervisor", async () => {
  const testSupervisorId = supervisionOfficerSupervisorsFixture[0].externalId;
  expect(store.officersBySupervisor.has(testSupervisorId)).toBeFalse();

  await expect(
    flowResult(store.hydrateOfficersForSupervisor(testSupervisorId))
  ).resolves.not.toThrow();

  expect(store.officersBySupervisor.get(testSupervisorId)).toEqual(
    expect.arrayContaining(supervisionOfficerFixture.slice(0, 2))
  );
});

test("current user record for supervisor except offline mode", () => {
  jest
    .spyOn(store.outliersStore.rootStore.userStore, "userAppMetadata", "get")
    .mockReturnValue({
      role: "supervision_staff",
      externalId: "abc123",
      district: "District One",
      stateCode: "us_mi",
    });

  if (isOfflineMode()) expect(store.currentSupervisorUser).toBeUndefined();

  expect(store.currentSupervisorUser).toBeDefined();
  expect(store.currentSupervisorUser).toMatchInlineSnapshot(`
    Object {
      "displayName": "",
      "externalId": "abc123",
      "fullName": Object {},
      "supervisionDistrict": "District One",
    }
  `);
});

test("current user record requires external ID", () => {
  jest
    .spyOn(store.outliersStore.rootStore.userStore, "userAppMetadata", "get")
    .mockReturnValue({
      role: "supervision_staff",
      district: "District One",
      stateCode: "us_mi",
    });

  expect(store.currentSupervisorUser).toBeUndefined();
});

test("no current user record for non-supervisor", () => {
  jest
    .spyOn(store.outliersStore.rootStore.userStore, "userAppMetadata", "get")
    .mockReturnValue({
      role: "leadership_role",
      externalId: "abc123",
      stateCode: "us_mi",
    });

  expect(store.currentSupervisorUser).toBeUndefined();
});

test("hydrate supervisors list with current user", () => {
  jest
    .spyOn(store.outliersStore.rootStore.userStore, "userAppMetadata", "get")
    .mockReturnValue({
      role: "supervision_staff",
      externalId: "abc123",
      district: "District One",
      stateCode: "us_mi",
    });

  expect(store.supervisionOfficerSupervisors).toBeDefined();
  expect(store.supervisionOfficerSupervisors).toMatchInlineSnapshot(`
    Array [
      Object {
        "displayName": "",
        "externalId": "abc123",
        "fullName": Object {},
        "supervisionDistrict": "District One",
      },
    ]
  `);
});

test("current supervisor user does not hydrate via API", async () => {
  jest
    .spyOn(store.outliersStore.rootStore.userStore, "userAppMetadata", "get")
    .mockReturnValue({
      role: "supervision_staff",
      externalId: "abc123",
      district: "District One",
      stateCode: "us_mi",
    });

  jest.spyOn(store.outliersStore.apiClient, "supervisionOfficerSupervisors");

  await expect(
    flowResult(store.hydrateSupervisionOfficerSupervisors())
  ).toResolve();
  expect(
    store.outliersStore.apiClient.supervisionOfficerSupervisors
  ).not.toHaveBeenCalled();
});

test("look up supervisor by ID", async () => {
  await flowResult(store.hydrateSupervisionOfficerSupervisors());

  const testSupervisor = supervisionOfficerSupervisorsFixture[0];
  expect(store.supervisionOfficerSupervisor(testSupervisor.externalId)).toEqual(
    testSupervisor
  );
});

test("hydrate supervisionOfficerMetricEvents", async () => {
  const testOfficer = supervisionOfficerFixture[1].externalId;
  const testMetric = supervisionOfficerFixture[2].outlierMetrics[0].metricId;

  function getTestEvents() {
    return store.metricEventsByOfficerAndMetricId
      .get(testOfficer)
      ?.get(testMetric);
  }

  expect(getTestEvents()).toBeUndefined();

  await flowResult(
    store.hydrateMetricEventsForOfficer(testOfficer, testMetric)
  );

  expect(getTestEvents()).toBeDefined();
  expect(getTestEvents()).toMatchSnapshot();
});
