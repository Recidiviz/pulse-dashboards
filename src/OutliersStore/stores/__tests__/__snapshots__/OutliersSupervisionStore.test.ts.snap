// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hydrate supervisionOfficerMetricEvents 1`] = `
Array [
  Object {
    "clientId": "985771",
    "clientName": Object {
      "givenNames": "Nicholas",
      "surname": "Rhodes",
    },
    "eventDate": 2022-12-24T00:00:00.000Z,
    "metricId": "absconsions_bench_warrants",
  },
  Object {
    "clientId": "326160",
    "clientName": Object {
      "givenNames": "Ethel",
      "surname": "Leonard",
    },
    "eventDate": 2022-12-21T00:00:00.000Z,
    "metricId": "absconsions_bench_warrants",
  },
  Object {
    "clientId": "931890",
    "clientName": Object {
      "givenNames": "Ricardo",
      "surname": "Wood",
    },
    "eventDate": 2022-12-24T00:00:00.000Z,
    "metricId": "absconsions_bench_warrants",
  },
  Object {
    "clientId": "930441",
    "clientName": Object {
      "givenNames": "Stanley",
      "surname": "Maxwell",
    },
    "eventDate": 2022-10-24T00:00:00.000Z,
    "metricId": "absconsions_bench_warrants",
  },
  Object {
    "clientId": "564940",
    "clientName": Object {
      "givenNames": "Miguel",
      "surname": "Haynes",
    },
    "eventDate": 2022-09-05T00:00:00.000Z,
    "metricId": "absconsions_bench_warrants",
  },
  Object {
    "clientId": "235261",
    "clientName": Object {
      "givenNames": "Rosie",
      "surname": "Luna",
    },
    "eventDate": 2023-05-06T00:00:00.000Z,
    "metricId": "absconsions_bench_warrants",
  },
  Object {
    "clientId": "531407",
    "clientName": Object {
      "givenNames": "Glenn",
      "surname": "Francis",
    },
    "eventDate": 2023-04-07T00:00:00.000Z,
    "metricId": "absconsions_bench_warrants",
  },
  Object {
    "clientId": "436502",
    "clientName": Object {
      "givenNames": "Steve",
      "surname": "Schwartz",
    },
    "eventDate": 2023-02-15T00:00:00.000Z,
    "metricId": "absconsions_bench_warrants",
  },
  Object {
    "clientId": "339710",
    "clientName": Object {
      "givenNames": "Bertha",
      "surname": "Figueroa",
    },
    "eventDate": 2023-05-14T00:00:00.000Z,
    "metricId": "absconsions_bench_warrants",
  },
]
`;
